### ISSUE TITLE
Implement human-readable quote file format

### ISSUE THREAD

**Issue by okitec** [AhiruSoftwareCompany/vaq#1]
Since multiline quotes are awdward in the current format, I propose a new one in which the quotes look like mails in an mbox-like format:

```
id: 1
date: 2018-12-06

Dozent: Keine Sorge, Sie sind nicht betrunken, das ist so unscharf.
---
id: 2
date: 2018-12-06

Der Server hier ist auf einem Auge blind. Also mindestens.
 ---
id: 4
date: 2018-12-13

Student: Ah, das hab ich garnicht mitbekommen.
Dozent: Das haben wir in der letzten Vorlesung gemacht.
Student: Da hab ich wohl gerade einen kurzen Powernap gemacht. 
```
This file is never modified by the program. Every entry consists of a header, the separating blank line and the body. Entries are separated but not terminated by three dashes. More header key-value pairs may be added later; vaq should ignore those it doesn't recognise. For parsing them, something like `http_parse_headers` can be used.

The relation `ID->rating` is stored in a different file and is kept up-to-date by the program. Every line is a (ID, rating) pair. Only pairs in which the rating is nonzero need to be stored. When an ID is apparently unused, but the rating is nonzero, it should be kept until explicit garbage collection (since the deletion
might have been a mistake).

```
1 2
2 -1
4 2
```

**Comment by leletec** [comment]
Two questions about the protocol:
- Should the client know / display the current voting on the quote it's showing?
- Should the client send the server the new absolute rating or just a relative one (i.e. {-1, +1})?

**Comment by leletec** [comment]
Current state:
- The client does not display or know the current rating of a given quote (the server always sends ```0``` as a default.
- As a result of that, the server receives a relative change in rating.
- The logic for the ```ratings``` file is implemented as follows (all ```PUT /quote```):
    * If the ```id``` is not already present in the file, it creates a new line with the given index and rating.
    * If the ```id``` is present, it adds the relativ change to the current rating.
    * If the ```id``` is present more than once, the duplicate entries get a ```-1 ``` prepended to the line, marking them invalid so that they can't be found again (```-1``` is an invalid id for a quote). For the sake of safety, they don't just get deleted.
- The changed file gets written as a whole.
- If the ```id``` of the received quote is ```-1```, the ```html_response_code``` will be ```400```, in all other cases ```200```.

**Comment by okitec** [comment]
I think the client should display the rating of a quote, maybe only after voting though. Reddit and SO votes are also public, after all. The client should tell the client when an upvote or downvote happens. We can't have the client just tell the server the new rating.

At the moment, there is no way to stop someone from upvoting again and again (no cookies/session key). For that to work, we wouldn't store the rating, we would store the relation (QuoteID,User,up/down) with the primary key (QuoteID,User) and store a list of users also.

**Comment by leletec** [comment]
> I think the client should display the rating of a quote, maybe only after voting though. Reddit and SO votes are also public, after all. The client should tell the client when an upvote or downvote happens. We can't have the client just tell the server the new rating.

Okay, not a big deal.

> At the moment, there is no way to stop someone from upvoting again and again (no cookies/session key). For that to work, we wouldn't store the rating, we would store the relation (QuoteID,User,up/down) with the primary key (QuoteID,User) and store a list of users also.

@okitec 
Would you implement it with a registration / login system?
Because since we use the concept of a REST-server, we can't save sessions etc. on the server and everthing on the client can be manipulated.
Sending your (encrypted) login credentials with every vote should fix that, but the question is if it is worth the additional effort, espacially keeping the different ```data/*``` files consistent.

**Comment by okitec** [comment]
We are not really fussed about a manipulative client, really. Apart from changing a quote's rating, there's not much that client could do. Just store a user key as a cookie.

**Comment by leletec** [comment]
In that case, the client might as well create and check for the cookie (with an expiration date) itself.

On another note: What do you think about changing the call for updating the rating from ```PUT /quote body={quote as JSON}``` to ```PUT /quote/{id} body={rating}```, since we don't need to transfer the whole quote-object in this direction?

**Comment by okitec** [comment]
Re: cookie, do what you think is best. I honestly don't know much about vaq at the moment.

Re: PUT, how about `PUT /q/<id>/up` and `PUT /q/<id>/down` with the user ID as body?

**Comment by leletec** [comment]
The new syntax to save a rating is now `PUT /quote/{id} {vote}`with rating being either -1 or 1 and id not being -1.

Example: `PUT /quote/2 1` will increment the rating of the the quote with index 2 by one.

**Comment by leletec** [comment]
I don't see the need for a user ID right now, since we won't use user accounts and limit the voting-rate with cookies.

**Comment by okitec** [comment]
With User ID, I meant the identifier saved in the cookie, nothing more.

Also I would disallow negative IDs in general.

**Comment by leletec** [comment]
Should a given client only be unable to vote on the same quote for a given time, or should he be completely excluded from voting?

**Comment by leletec** [comment]
Another easy way to prevent duplicate votes would be to just load a new quote after you voted on one.

**Comment by okitec** [comment]
You should be able to look at quotes you've voted on previously, and you shouldn't be able to vote again except perhaps to retract or change your vote. Compare YT likes/dislikes, Reddit, any old voting system.

**Comment by leletec** [comment]
In that case you would need account management though, because otherwise you can't make the data persistent over multiple sessions / machines.

**Comment by okitec** [comment]
The data stays over multiple sessions as long as the cookie exists. Yes, you can delete the cookie and vote again, use another device, another browser. But the question is: who would gain anything by manipulating quote votes? The cookie is merely a safeguard against **accidentally** upvoting/downvoting something more than once. Accounts would be too cumbersome.

Look at Strawpoll and other such systems: they also just use a cookie and rely on the vote participants not being jerks.

**Comment by leletec** [comment]
So I see some possibilities of a cookie based system:
 1. The client saves a cookie with a randomly generated (?) number. The server then saves a table user_id => array( id => voting ).
 2. The client saves the voting information in the cookie itself (depending on how much you can put in there, in different cookies).

Working with $_SESSION variables would be nice, but I can't find a way to easily recreate the session with just the session_id, after the session gets destroyed (client closes the browser).

**Comment by okitec** [comment]
I'd prefer the first option. It's the server's job to store the table.

**Comment by leletec** [comment]
Soo, that commit should do that :smile: .

**Comment by okitec** [comment]
What a hefty commit! Now the only thing missing is the parsing of the quote format before we can close this issue, right? :+1:

On an unrelated note, I noticed some of your commit messages being so long because of an explanation in them. I'd suggest putting the explanation and other details in the body of the commit message instead of cramming everything in the headline.

**Comment by leletec** [comment]
Re parser: Do you mean `zitatliste` -> `quotes`? I made a new issue for that task, because it is not necessaryly that related to the PHP server.

Re commits: I noticed that too, will try to follow your advice in the future.

**Comment by okitec** [comment]
Nevermind, I had thought that the parsing of the quote file format wasn't implemented yet, but it is. Didn't look at all the code and just noticed. So, will you close this issue now?

**Comment by leletec** [comment]
Yup, should be all done for now.
